### PR TITLE
refactoring the Release script

### PR DESCRIPTION
**Pull request by andreeap**
This is from Friday, still learning.

### CODE REVIEW

File: lib/MetaCPAN/Script/Release.pm [hunk 1]
```diff
@@ -466,7 +404,7 @@ sub import_tarball {
         }
     }
     if (@provides) {
-        $release->provides( [ sort @provides ] );
+        $release->provides( \@provides );
```
Comment (oalders) [inline lib/MetaCPAN/Script/Release.pm]: Why did you remove the `sort`?

File: lib/MetaCPAN/Model/Tarball.pm [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+    default => '/tmp',
+);
+
+has date => (
+    is  => 'ro',
+    isa => 'DateTime',
+);
+
+has index => (
+    is         => 'ro',
```
Comment (oalders) [inline lib/MetaCPAN/Model/Tarball.pm]: All of these attributes should include `isa`.

File: lib/MetaCPAN/Model/Tarball.pm [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+    lazy_build => 1,
```
Comment (oalders) [inline lib/MetaCPAN/Model/Tarball.pm]: Other code may use `lazy_build`, but that's not a practice that's encouraged.

```
lazy => 1,
builder => '_build_foo',
```

File: lib/MetaCPAN/Model/Tarball.pm [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+has release => (
+    is         => 'ro',
+    lazy_build => 1,
+);
+
+sub _build_files {
+    my $self = shift;
+    log_info {"Processing $self->tarball"};
+
+    my $at = Archive::Any->new( $self->tarball );
```
Comment (oalders) [inline lib/MetaCPAN/Model/Tarball.pm]: What does `$at` mean?  Variable names should be descriptive and unambiguous.

File: lib/MetaCPAN/Model/Tarball.pm [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+        = dir( File::Temp::tempdir( CLEANUP => 0, DIR => $self->base_dir ) );
+    log_error {"$self is being impolite"} if $at->is_impolite;
+
+    log_error {"$self is being naughty"} if $at->is_naughty;
+
+    log_debug {"Extracting archive to filesystem"};
+    $at->extract($tmpdir);
+
+    my @files;
+    my @list = $at->files;
```
Comment (oalders) [inline lib/MetaCPAN/Model/Tarball.pm]: `@list` is pretty non-descriptive too, but you only use it once, so you don't need to create the variable.

```
log_debug { 'Indexing ', scalar $at->files, " files" };
```

File: lib/MetaCPAN/Script/Release.pm (hunk 1 above)
Comment (andreeap) [inline lib/MetaCPAN/Script/Release.pm]: In the new version I did not, this why I said I messed up with git, I solved some merge conflicts and deleted some things that shouldn't have been deleted. The sort stays.

File: lib/MetaCPAN/Model/Tarball.pm [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+    if ( -l $self ) {
+        my $syml = readlink $self;
+        return 1 if ( !-e $syml && !-l $syml );
+    }
+    return 0;
+}
+
+sub is_broken_file {
+    my $self = shift;
+    return 1 if ( -p $self || !-e $self );
```
Comment (oalders) [inline lib/MetaCPAN/Model/Tarball.pm]: You can do this in one line:

```perl
return ( -p $self || !-e $self );
```

File: lib/MetaCPAN/Script/Release.pm [hunk 7]
```diff
@@ -202,36 +202,25 @@ sub run {
 
 sub import_tarball {
     my ( $self, $tarball ) = @_;
+
+    $tarball = MetaCPAN::Model::Tarball->new( tarball => $tarball );
```
Comment (oalders) [inline lib/MetaCPAN/Script/Release.pm]: `my $tarball`

File: lib/MetaCPAN/Model/Tarball.pm (hunk 2 above)
Comment (andreeap) [inline lib/MetaCPAN/Model/Tarball.pm]: I am still figuring which one what is, but again, there are not too many attributes? When I add a new file, I need too much information and I don't know how else to carry that.

File: lib/MetaCPAN/Script/Release.pm [hunk 8]
```diff
@@ -202,36 +202,25 @@ sub run {
...
-    log_info {"Processing $tarball"};
-
-    # load Archive::Any in the child due to bugs in MMagic and MIME::Types
-    require Archive::Any;
-    my $at = Archive::Any->new($tarball);
-    my $tmpdir
-        = dir( File::Temp::tempdir( CLEANUP => 0, DIR => $self->base_dir ) );
-
-    log_error {"$tarball is being impolite"} if $at->is_impolite;
+    $tarball->author = $author;
```
Comment (oalders) [inline lib/MetaCPAN/Script/Release.pm]: What are you trying to do here?  This?

```perl
$tarball->author( $author );
```

File: lib/MetaCPAN/Model/Tarball.pm (hunk 4 above)
Comment (andreeap) [inline lib/MetaCPAN/Model/Tarball.pm]: I will named it $archive.

File: lib/MetaCPAN/Script/Release.pm [hunk 9]
```diff
@@ -202,36 +202,25 @@ sub run {
...
+    $tarball->name   = $name;
 
-    # TODO: add release to the index with status => 'broken' and move along
-    log_error {"$tarball is being naughty"} if $at->is_naughty;
-
-    log_debug {"Extracting archive to filesystem"};
-    $at->extract($tmpdir);
-
-    my $date    = DateTime->from_epoch( epoch => $tarball->stat->mtime );
+    $tarball->date = DateTime->from_epoch( epoch => $tarball->stat->mtime );
```
Comment (oalders) [inline lib/MetaCPAN/Script/Release.pm]: See above.  Moose attributes should be set via accessors.

File: lib/MetaCPAN/Model/Tarball.pm (hunk 3 above)
Comment (andreeap) [inline lib/MetaCPAN/Model/Tarball.pm]: I though it was a good thing, but now I know.

File: lib/MetaCPAN/Script/Release.pm (hunk 8 above)
Comment (oalders) [inline lib/MetaCPAN/Script/Release.pm]: This actually won't work either since the attribute is `ro`.  If you the object needs this attribute, you should pass it when the object is created, via `new( author => 'foo', ... )`

Comment (andreeap) [inline lib/MetaCPAN/Script/Release.pm]: Yes

**Comment by andreeap** [comment]
I have not put 'isa' everywhere because I am not sure how to add their type.

File: cpanfile [hunk 10]
```diff
@@ -95,6 +95,7 @@ requires 'MooseX::ClassAttribute';
 requires 'MooseX::Getopt';
 requires 'MooseX::Getopt::Dashes';
 requires 'MooseX::Getopt::OptionTypeMap';
+requires 'MooseX::StrictConstructor';
```
Comment (oalders) [inline cpanfile]: When you add a new module, make sure you run carton and then commit the changes to cpanfile.snapshot so that we can mimic your installed modules in production.

File: lib/MetaCPAN/Model/Tarball.pm [hunk 11]
```diff
@@ -0,0 +1,167 @@
+package MetaCPAN::Model::Tarball;
+
+use Archive::Any;
+use CPAN::DistnameInfo ();
+use DateTime           ();
+use DDP;
+use Log::Contextual qw( :log :dlog );
+use MetaCPAN::Types qw(ArrayRef File Dir);
```
Comment (oalders) [inline lib/MetaCPAN/Model/Tarball.pm]: `use MetaCPAN::Types qw(ArrayRef Dir File HashRef Str);`

Then you can do `isa => Str` (no quotes) etc.

File: lib/MetaCPAN/Model/Tarball.pm [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+                my $s = $child->stat;
+                +{ map { $_ => $s->$_ } qw(mode uid gid size mtime) };
+            };
+            return if _is_broken_symlink($File::Find::name);
+            return if _is_broken_file($File::Find::name);
+            return if ( $relative eq q{.} );
+            ( my $fpath = "$relative" ) =~ s/^.*?\///;
+            my $fname = $fpath;
+            $child->is_dir
+                ? $fname =~ s/^(.*\/)?(.+?)\/?$/$2/
```
Comment (oalders) [inline lib/MetaCPAN/Model/Tarball.pm]: We should use a more descriptive variable name, like `$filename` rather than `$fname`.

File: lib/MetaCPAN/Model/Tarball.pm [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+                    release      => $self->name,
+                    date         => $self->date,
+                    distribution => $self->distribution,
+                    author       => $self->author,
+                    local_path   => $child,
+                    path         => $fpath,
+                    version      => $self->version,
+                    stat         => $self->status,
+                    maturity     => $self->maturity,
+                    status       => $self->release->status,
```
Comment (oalders) [inline lib/MetaCPAN/Model/Tarball.pm]: If you alpha sort the hash keys, this becomes easier to read and duplicate keys become easier to spot.

File: lib/MetaCPAN/Model/Tarball.pm [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+    return \@files;
+}
+
+sub _is_broken_symlink {
+    my $self = shift;
+    if ( -l $self ) {
+        my $syml = readlink $self;
+        return 1 if ( !-e $syml && !-l $syml );
+    }
+    return 0;
```
Comment (oalders) [inline lib/MetaCPAN/Model/Tarball.pm]:
```perl
sub _is_broken_symlink {
    my $self = shift;
    my $filename = shift;
    return 0 if ( -l $filename );

    my $syml = readlink $filename;
    return ( !-e $syml && !-l $syml );
}
```

File: lib/MetaCPAN/Model/Tarball.pm [hunk 15]
```diff
@@ -0,0 +1,167 @@
...
+}
+
+sub _is_broken_file {
+    my $self = shift;
+    return ( -p $self || !-e $self );
+}
+
```
Comment (oalders) [inline lib/MetaCPAN/Model/Tarball.pm]:
```perl
sub _is_broken_file {
    my $self = shift;
    my $filename = shift;
    return ( -p $filename || !-e $filename );
}
```

File: lib/MetaCPAN/Script/Release.pm [hunk 16]
```diff
@@ -201,30 +202,16 @@ sub run {
 }
 
 sub import_tarball {
-    my ( $self, $tarball ) = @_;
+    my ( $self, $tar ) = @_;
     my $cpan = $self->index;
 
-    $tarball = Path::Class::File->new($tarball);
-    my $d = CPAN::DistnameInfo->new($tarball);
+    $tar = Path::Class::File->new($tar);
```
Comment (oalders) [inline lib/MetaCPAN/Script/Release.pm]: You're redefining `$tar` here.  Either it wasn't actually an argument to `import_tarball` or you shouldn't be redefining it.

File: lib/MetaCPAN/Script/Release.pm [hunk 17]
```diff
@@ -201,30 +202,16 @@ sub run {
 }
 
 sub import_tarball {
-    my ( $self, $tarball ) = @_;
+    my ( $self, $tar ) = @_;
```
Comment (oalders) [inline lib/MetaCPAN/Script/Release.pm]: `$tar` doesn't tell me what's in this variable.

File: lib/MetaCPAN/Script/Release.pm [hunk 18]
```diff
@@ -280,60 +267,24 @@ sub import_tarball {
...
+        author       => $author,
+        name         => $name,
+        date         => $date,
+        metadata     => $meta,
+        index        => $cpan,
+        status       => $stat,
+        distribution => $d->dist,
+        version      => $d->version,
+        maturity     => $d->maturity,
+        bulk         => $bulk,
```
Comment (oalders) [inline lib/MetaCPAN/Script/Release.pm]: These keys should be alpha sorted as well.

**Comment by oalders** [comment]
@andreeap what's the status of this pull request?

**Comment by schwern** [comment]
I'm done refactoring.  It's far from perfect, I was very rote and conservative and leaned heavily on the existing tests.  There's some attribute name inconsistencies between Document::Release, Script::Release and Model::Release it would be nice to resolve.

This is enough to start hooking in Gitpan.

**Comment by oalders** [comment]
I don't see anything here which causes me any concerns.  Also, I feel like this is a nice cleanup over the existing code.  It's way less confusing now that things are more compartmentalized.  I'd like to leave this open for a day or two while we get some more feedback, but I also don't want to leave this open for very long.  If there's no dissent, I say we merge this on Monday (Feb 23).

@andreeap how much have you tested this?  If you haven't done a full re-index via this branch, now would be a good time to kick that off.

**Comment by andreeap** [comment]
I have tested this on my VM and everything seemed fine at indexing, but I was indexing more than one release and when I tried with just one I noticed that for every release I will get 3 x ```D release: Extracting archive to filesystem``` which means it will try to extract the archive for so many times. This should really happen? 
I started a full re-index and I got a few of these(but from what I have seen are just the usual errors and because of the changes some sound different)
<pre><code>
2015/02/20 10:26:06 I release: Processing /home/metacpan/CPAN/authors/id/R/RI/RIK/NIS-a2.tar.Z
No handler available for type 'application/x-compress' at /home/metacpan/carton/metacpan-api/lib/perl5/Archive/Any.pm line 52.
2015/02/20 10:26:06 F release: Attribute (_extractor) does not pass the type constraint because: Validation failed for 'Archive::Any' with value undef at reader MetaCPAN::Model::Archive::_extractor (defined at /home/metacpan/metacpan-api/bin/../lib/MetaCPAN/Model/Archive.pm line 69) line 8.
	MetaCPAN::Model::Archive::_extractor('MetaCPAN::Model::Archive=HASH(0x60bbf50)') called at /home/metacpan/carton/metacpan-api/lib/perl5/x86_64-linux/Moose/Meta/Method/Delegation.pm line 92
	MetaCPAN::Model::Archive::is_impolite('MetaCPAN::Model::Archive=HASH(0x60bbf50)') called at /home/metacpan/metacpan-api/bin/../lib/MetaCPAN/Model/Release.pm line 128
	MetaCPAN::Model::Release::_build_archive('MetaCPAN::Model::Release=HASH(0xfafe990)') called at reader MetaCPAN::Model::Release::archive (defined at /home/metacpan/metacpan-api/bin/../lib/MetaCPAN/Model/Release.pm line 21) line 7
	MetaCPAN::Model::Release::archive('MetaCPAN::Model::Release=HASH(0xfafe990)') called at /home/metacpan/metacpan-api/bin/../lib/MetaCPAN/Model/Release.pm line 216
	MetaCPAN::Model::Release::__ANON__() called at /home/metacpan/carton/metacpan-api/lib/perl5/Log/Contextual/Router.pm line 138
	Log::Contextual::Router::handle_log_request('Log::Contextual::Router=HASH(0x3992b10)', 'exporter', 'Log::Contextual', 'caller_package', 'MetaCPAN::Model::Release', 'caller_level', 1, 'message_level', 'debug', 'message_sub', 'CODE(0xfafec90)', 'message_args', 'ARRAY(0x5186bb8)') called at /home/metacpan/carton/metacpan-api/lib/perl5/Log/Contextual.pm line 130
	Log::Contextual::__ANON__('CODE(0xfafec90)') called at /home/metacpan/metacpan-api/bin/../lib/MetaCPAN/Model/Release.pm line 216
	MetaCPAN::Model::Release::_build_files('MetaCPAN::Model::Release=HASH(0xfafe990)') called at reader MetaCPAN::Model::Release::files (defined at /home/metacpan/metacpan-api/bin/../lib/MetaCPAN/Model/Release.pm line 63) line 7
	MetaCPAN::Model::Release::files('MetaCPAN::Model::Release=HASH(0xfafe990)') called at /home/metacpan/metacpan-api/bin/../lib/MetaCPAN/Script/Release.pm line 196
	MetaCPAN::Script::Release::import_archive('MetaCPAN::Script::Release=HASH(0x5eef620)', '/home/metacpan/CPAN/authors/id/R/RI/RIK/NIS-a2.tar.Z') called at /home/metacpan/metacpan-api/bin/../lib/MetaCPAN/Script/Release.pm line 163
	MetaCPAN::Script::Release::try {...} () called at /home/metacpan/carton/metacpan-api/lib/perl5/Try/Tiny.pm line 81
	eval {...} called at /home/metacpan/carton/metacpan-api/lib/perl5/Try/Tiny.pm line 72
	Try::Tiny::try('CODE(0xfb01e00)', 'Try::Tiny::Catch=REF(0x93609b0)') called at /home/metacpan/metacpan-api/bin/../lib/MetaCPAN/Script/Release.pm line 166
	MetaCPAN::Script::Release::run('MetaCPAN::Script::Release=HASH(0x5eef620)') called at /home/metacpan/carton/metacpan-api/lib/perl5/x86_64-linux/Class/MOP/Method/Wrapped.pm line 50
	Class::MOP::Method::Wrapped::__ANON__('MetaCPAN::Script::Release=HASH(0x5eef620)') called at /home/metacpan/carton/metacpan-api/lib/perl5/x86_64-linux/Class/MOP/Method/Wrapped.pm line 91
	MetaCPAN::Script::Release::run('MetaCPAN::Script::Release=HASH(0x5eef620)') called at /home/metacpan/metacpan-api/bin/../lib/MetaCPAN/Script/Runner.pm line 32
	MetaCPAN::Script::Runner::run('MetaCPAN::Script::Runner') called at /home/metacpan/metacpan-api/bin/metacpan line 21

</code></pre>

**Comment by schwern** [comment]
No, it's only being extracted once, but it's being asked to be extracted multiple times.

The logging for extract is in `Model::Release->extract()`, but the actual extraction is done in `Model::Archive::extract()`.  `Model::Archive->extract()` will only extract once, thereafter it will do nothing.  I avoided putting the log message in Module::Archive because it would make initializing Module::Archive more involved and make unit testing harder.

It might make more sense to add a log initialization function to a test helper module.

**Comment by oalders** [comment]
I'm going to merge this now, since overly verbose logging seems like a secondary concern.

**Comment by schwern** [comment]
Sorry, I got distracted.
